File: surface/widgets/compose/components/widgets/form-controls/ToggleSwitch.kt
// ============================================================
// Clef Surface Compose Widget — ToggleSwitch
//
// Binary on/off toggle control. Renders a Material 3 Switch
// with an optional label. Maps the toggle-switch.widget
// anatomy (root, input, control, thumb, label) to Compose
// Switch with Row-based label alignment.
// ============================================================

package clef.surface.compose.components.widgets.formcontrols

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.selection.toggleable
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.dp

// --------------- Component ---------------

@Composable
fun ToggleSwitch(
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
    label: String? = null,
    enabled: Boolean = true,
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier.toggleable(
            value = checked,
            enabled = enabled,
            role = Role.Switch,
            onValueChange = onCheckedChange,
        ),
    ) {
        Switch(
            checked = checked,
            onCheckedChange = null, // handled by toggleable
            enabled = enabled,
        )

        if (label != null) {
            Spacer(modifier = Modifier.width(12.dp))
            Text(
                text = label,
                style = MaterialTheme.typography.bodyMedium,
                color = if (enabled) {
                    MaterialTheme.colorScheme.onSurface
                } else {
                    MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
                },
            )
        }
    }
}

File: surface/widgets/gtk/components/widgets/form-controls/ToggleSwitch.ts
// ============================================================
// Clef Surface GTK Widget — ToggleSwitch
//
// Binary on/off toggle control. Renders a Gtk.Switch with an
// optional label. Maps the toggle-switch.widget anatomy (root,
// input, control, thumb, label) to Gtk.Switch with label.
// ============================================================

import Gtk from 'gi://Gtk?version=4.0';

// --------------- Props ---------------

export interface ToggleSwitchProps {
  checked?: boolean;
  label?: string | null;
  disabled?: boolean;
  onCheckedChange?: (checked: boolean) => void;
}

// --------------- Component ---------------

/**
 * Creates a GTK4 Switch widget for binary on/off toggling with
 * an optional label.
 */
export function createToggleSwitch(props: ToggleSwitchProps = {}): Gtk.Widget {
  const {
    checked = false,
    label = null,
    disabled = false,
    onCheckedChange,
  } = props;

  const gtkSwitch = new Gtk.Switch({
    active: checked,
  });

  gtkSwitch.set_sensitive(!disabled);

  if (onCheckedChange) {
    gtkSwitch.connect('notify::active', () => {
      onCheckedChange(gtkSwitch.get_active());
    });
  }

  if (label) {
    const box = new Gtk.Box({
      orientation: Gtk.Orientation.HORIZONTAL,
      spacing: 12,
    });
    box.append(gtkSwitch);
    box.append(new Gtk.Label({ label }));
    return box;
  }

  return gtkSwitch;
}
